internal/runner/command.go

```diff
@@ -0,0 +1,340 @@
+
+	// pty and tty as pseud-terminal primary and secondary.
+	// Might be nil if not allocating a pseudo-terminal.
+	pty *os.File
+	tty *os.File
+
+	envDir string
+
+	cmd *exec.Cmd
+
+	done chan struct{}
+	wg   sync.WaitGroup
+	mu   sync.Mutex
+	err  error
+
+	logger *zap.Logger
+}
+
+func (c *command) seterr(err error) {
+	if err == nil {
+		return
+	}
+
+	c.mu.Lock()
+	if c.err == nil {
+		c.err = err
+	}
+	c.mu.Unlock()
+}
+
+type commandConfig struct {
+	ProgramName string   // a path to shell or a name, for example: "/usr/local/bin/bash", "bash"
+	Args        []string // args passed to the program
+	Directory   string
+	Envs        []string
+	Tty         bool // if true, a pseudo-terminal is allocated
+
+	IsShell  bool // if true then Commands or Scripts is passed to shell as "-c" argument's value
+	Commands []string
+	Script   string
+
+	Input []byte // initial input data passed immediately to the command.
+}
+
+func newCommand(
+	cfg *commandConfig,
+	logger *zap.Logger,
+) (*command, error) {
+	programPath, err := exec.LookPath(cfg.ProgramName)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	directory := cfg.Directory
+	if directory == "" {
+		var err error
+		directory, err = os.Getwd()
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+	}
+
+	var (
+		extraArgs    []string
+		envStorePath string
+	)
+
+	if cfg.IsShell && (len(cfg.Commands) > 0 || cfg.Script != "") {
+		var err error
+		envStorePath, err = os.MkdirTemp("", "")
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+
+		var script strings.Builder
+
+		_, _ = script.WriteString("env > " + filepath.Join(envStorePath, envStartFileName) + "\n")
+
+		if len(cfg.Commands) > 0 {
+			_, _ = script.WriteString(
+				prepareScriptFromCommands(cfg.Commands),
+			)
+		} else if cfg.Script != "" {
+			_, _ = script.WriteString(
+				prepareScript(cfg.Script),
+			)
+		}
+
+		_, _ = script.WriteString("env > " + filepath.Join(envStorePath, envEndFileName) + "\n")
+
+		extraArgs = []string{"-c", script.String()}
+	}
+
+	cmd := &command{
+		ProgramPath: programPath,
+		Args:        append(cfg.Args, extraArgs...),
+		Directory:   directory,
+		Envs:        cfg.Envs,
+		cfg:         cfg,
+		envDir:      envStorePath,
+		done:        make(chan struct{}),
+		logger:      logger,
+	}
+
+	if cfg.Tty {
+		var err error
+		cmd.pty, cmd.tty, err = pty.Open()
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+
+		cmd.Stdin = cmd.pty
+		if len(cfg.Input) > 0 {
+			_, err := io.Copy(cmd.Stdin, bytes.NewReader(cfg.Input))
+			if err != nil {
+				cmd.preWaitCleanup()
+				return nil, errors.WithMessage(err, "failed to write initial input")
+			}
+		}
+
+		// stdout is read from pty. stderr is unused because
+		// it can't be distinguished from pty.
+		cmd.Stdout = rbuffer.NewRingBuffer(4096)
+		cmd.Stderr = rbuffer.NewRingBuffer(0)
+	} else {
+		cmd.Stdin = &safeBuffer{buf: bytes.NewBuffer(cfg.Input)}
+		cmd.Stdout = rbuffer.NewRingBuffer(4096)
```

Comment:
Why use a ring buffer here? I realize you're concerned about stress on the GC, but I feel like this is an extreme solution. I could imagine, for example, a large JSON payload easily exceeding `4kb`. Or, if many lines are sent as one `stdout` message, that could also easily exceed the limit.

I don't know a lot about go's GC, but is it going to stress the GC severely if we use a dynamic queue instead?

Reply:
It's not really about GC here, because `bytes.Buffer` implementation is pretty smart about it. It's about avoiding the buffer growing indefinitely. I agree that we probably should adjust the numbers, but it can wait until we start testing it. It can be parametrized in the create session request as well.